src/serializer/ResidualHeapSerializer.js

```diff
@@ -1161,6 +1221,83 @@ export class ResidualHeapSerializer {
     return false;
   }
 
+  processAdditionalFunctionValues(): Map<FunctionValue, Array<BabelNodeStatement>> {
+    let rewrittenAdditionalFunctions: Map<FunctionValue, Array<BabelNodeStatement>> = new Map();
+    let shouldEmitLog = !this.residualHeapValueIdentifiers.collectValToRefCountOnly;
+    let processAdditionalFunctionValuesFn = () => {
+      let additionalFVEffects = this.additionalFunctionValuesAndEffects;
+      if (additionalFVEffects) {
+        for (let [additionalFunctionValue, effects] of additionalFVEffects.entries()) {
+          let [r, g, ob, pb: Map<PropertyBinding, void | Descriptor>, co] = effects;
+          let nestedFunctions = new Set([...co].filter(object => object instanceof FunctionValue));
+          // result -- ignore TODO: return the result from the function somehow
+          // Generator -- visit all entries
+          // Bindings -- only need to serialize bindings if they're captured by some nested function ??
+          //          -- need to apply them and maybe need to revisit functions in ancestors to make sure
+          //          -- we don't overwrite anything they capture
+          //          -- TODO: deal with these properly
+          // PropertyBindings -- visit any property bindings that aren't to createdobjects
+          // CreatedObjects -- should take care of itself
+          this.realm.applyEffects([r, new Generator(this.realm), ob, pb, co]);
+          // Allows us to emit function declarations etc. inside of this additional
+          // function instead of adding them at global scope
+          // TODO: make sure this generator isn't getting mutated oddly
+          this.additionalFunctionValueNestedFunctions = ((nestedFunctions: any): Set<FunctionValue>);
+          let emitter = this.emitter;
+          let serializePropertiesAndBindings = () => {
+            for (let propertyBinding of pb.keys()) {
+              let binding: PropertyBinding = ((propertyBinding: any): PropertyBinding);
+              let object = binding.object;
+              // These should be in the generator so we can skip them
+              if (object.isPartialObject()) continue;
+              invariant(object instanceof ObjectValue);
+              if (co.has(object)) continue; // Created Object's binding
+              if (object.intrinsicName === "global") continue; // Avoid double-serialization
+              if (object.refuseSerialization) continue; // modification to internal state
+              this._emitProperty(object, binding.key, binding.descriptor, true);
+            }
+            for (let binding of ob.keys()) {
+              let old_value = ob.get(binding);
+              let new_value = binding.value;
+              invariant(old_value !== new_value);
+              if (old_value) {
+                let lhs = ((this.serializeValue(old_value): any): BabelNodeIdentifier);
+                invariant(t.isIdentifier(lhs));
+                emitter.emit(
```

Comment:
Redefining binding values like this seems wrong to me. If this does what I think it is intended to do, then the right way of doing it must go through the "referentialization" phase, the referentialized location needs to get updated. Values and locations where values are stored are not the same thing.